docs/pages/project/contributing/contributing-rego.md

```diff
@@ -0,0 +1,42 @@
+
+## Components of Meshery Policy Engine
+
+Rego policies are the declarative logic behind the policy engine. These policies define how the underlying engine understands relationships within Meshery Designs. Rego policies can be found [here](https://github.com/meshery/meshery/tree/master/server/meshmodel/kubernetes/v1.25.2/v1.0.0/policies) in the Meshery Github repository.
+
+Currently the Rego policies are invoked from the Go code in Meshery Server. This requires configuring the OPA context as seen [here](https://github.com/meshery/meshkit/blob/master/models/meshmodel/core/policies/rego_policy_relationship.go).
+
+The Meshery Policy Engine is invoked by calling the evaluation endpoint defined [here](https://github.com/meshery/meshery/blob/master/server/handlers/policy_relationship_handler.go). This endpoint is passed a Meshery Design in JSON which it passes to the Meshery Policy Engine for validation.
+
+## Working with Meshery Policy Engine
+
+Working with the Meshery Policy Engine can feel complex because of the large number of different relationships a Meshery Design might contain. Contributors must then be careful to test any changes thoroughly to avoid unexpected results. When working with the Rego policies themselves it makes testing and development easier to be able to test the policies directly without running the entire Meshery Server. When working on the policies keep the following in mind:
+
+- Since we will be running these policies using the OPA CLI directly without the Go bootstrap code we need to provide the necessary context for the policy engine. This means providing a data structure containing the valid Meshery Relationships the policy engine will validate against. **Currently there is no automated process for getting this relationship data.**
+- The input to the policy engine is a Meshery Design as a JSON data structure. When testing the Meshery Policy Engine you will need to provide this design as input. If you are working on a bug you will need to get the design from the issue or reporter.
+
+## Executing OPA 
+
+Now that you have the necessary prerequisites you can try running the Rego policies locally. From your local version of Meshery Server navigate to the policies directory:
+{% capture code_content %}cd server/meshmodel/kubernetes/v1.25.2/v1.0.0/policies{% endcapture %}
+{% include code.html code=code_content %}
+
+Assuming you have the design you are testing in the parent directory in a file named **input.json** and the relationship data in the parent directory in a file named **all_relationships.json** you can evaluate the design against the Rego policies using this command:
+{% capture code_content %}opa eval -i ../input.json -d relationships:../all_relationships.json -d . 'data.relationship_evaluation_policy.evaluate'{% endcapture %}
+{% include code.html code=code_content %}
+
```

Comment:
In addition to Regal as a linter, we are need of identifying and reusing a test harness from the OPA project. Is there such a test harness prominently available or should we ask one of OPA maintainers for recommendations?

Reply:
The testing tools provided from OPA seem to provide what would be needed for implementing tests including naming convention for test files, how to provide test data into the Rego context:

https://www.openpolicyagent.org/docs/latest/policy-testing

There is an assertion library to make tests for readable:

https://github.com/anderseknert/rego-test-assertions

And a Github Action to execute tests and generate a report:

https://github.com/masterpointio/github-action-opa-rego-test

Put all of that together and it seems to be a reasonable starting place.

Reply:
These are quite promising, I agree. The second item in the list, the assertion library, is from our friend Anders, which is great and gives credence to its usefulness. All three of these resources focus on testing at the unit test level with `_test` files and `opa test`, is that right?

I wonder if there are any libraries or actions that focus on functional tests in which we're able to verify that the output is consistent with the end user's expectations; simulate different design situations from the end-user's perspective.